### PR TITLE
Feature/s3 browser action additions

### PR DESCRIPTION
**Pull request by shruti0085**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ x] New feature (non-breaking change which adds functionality)

## Description

Added bucket and object actions for S3.

## Testing

All tests are unit tests.

## Screenshots (if appropriate)
Bucket Actions [Updated]
updatedBucketActions

Object Actions 
objectactions

Edge cases where certain actions are disabled
edgeCases

Open Action indexing fix
openaction

## Checklist

- [x ] I have read the **README** document
- [ x] I have read the **CONTRIBUTING** document
- [ x] Local run of `gradlew check` succeeds
- [ x] My code follows the code style of this project
- [ x] I have added tests to cover my changes
- [ x] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**

## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/bucketActions/CreateS3Bucket.kt [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+import com.intellij.openapi.actionSystem.AnActionEvent
+import com.intellij.openapi.actionSystem.LangDataKeys
+import icons.AwsIcons
+import software.amazon.awssdk.services.s3.S3Client
+import software.aws.toolkits.jetbrains.components.telemetry.AnActionWrapper
+import software.aws.toolkits.jetbrains.core.AwsClientManager
+import software.aws.toolkits.jetbrains.services.s3.CreateS3BucketDialog
+
+class CreateS3Bucket : AnActionWrapper("Create New S3 Bucket", null, AwsIcons.Actions.LAMBDA_FUNCTION_NEW) {
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/bucketActions/CreateS3Bucket.kt]: Can we get this as a `message` string id?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3TreeCellRenderer.kt [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    override fun getTreeCellRendererComponent(
+        tree: JTree?,
+        value: Any?,
+        sel: Boolean,
+        expanded: Boolean,
+        leaf: Boolean,
+        row: Int,
+        hasFocus: Boolean
+    ): Component {
+        val renderer = super.getTreeCellRendererComponent(tree, value, sel, expanded, leaf, row, hasFocus) as JLabel
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3TreeCellRenderer.kt]: Nit: this is really a `component` rather than a renderer.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3TreeCellRenderer.kt [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        if (value is DefaultMutableTreeNode) {
+            val userObject = value.userObject
+            if (userObject is S3KeyNode) {
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3TreeCellRenderer.kt]: I thought we had different representations of 'object' vs 'directory' as a base concept (extensions of `S3KeyNode`). If so this (and the nested if) could be replaced with a `when` block.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3ViewerProvider.kt [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    override fun accept(project: Project, file: VirtualFile) = file is S3VirtualBucket
+
+    override fun createEditor(project: Project, file: VirtualFile) = S3ViewerEditor(file as S3VirtualBucket)
+
+    override fun getPolicy() = FileEditorPolicy.HIDE_DEFAULT_EDITOR
+
+    override fun getEditorTypeId() = EDITOR_TYPE_ID
+
+    companion object {
+        const val EDITOR_TYPE_ID = "S3 Bucket Viewer"
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3ViewerProvider.kt]: Is this ID visible to the customer? If so it should be a `message`

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/bucketActions/CopyAction.kt [hunk 5]
```diff
@@ -0,0 +1,13 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+package software.aws.toolkits.jetbrains.services.s3.bucketActions
+
+import com.intellij.icons.AllIcons
+import com.intellij.openapi.project.DumbAware
+import software.aws.toolkits.jetbrains.core.explorer.SingleResourceNodeAction
+import software.aws.toolkits.jetbrains.core.explorer.nodes.AwsExplorerResourceNode
+
+abstract class CopyAction<in T : AwsExplorerResourceNode<*>>(text: String) : SingleResourceNodeAction<T>(text, icon = AllIcons.Actions.Copy),
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/bucketActions/CopyAction.kt]: I like the idea of an abstract copy action but we should put the logic for interacting with `CopyPasteManager` - the concrete implementation just needs to provide how to get a string from `T`

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/bucketEditor/S3ViewerPanel.java [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+import javax.swing.Action;
+import javax.swing.KeyStroke;
+import javax.swing.SwingConstants;
+import javax.swing.border.TitledBorder;
+import javax.swing.table.DefaultTableCellRenderer;
+import javax.swing.tree.DefaultMutableTreeNode;
+import java.awt.event.ActionEvent;
+
+@SuppressWarnings("unchecked")
+
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/bucketEditor/S3ViewerPanel.java]: Extra space

**Comment by shruti0085** [comment]
> Still working thru, some intial feedback done. We need all customer facing strings to be in `message`. It'd also be great to get some pictures (or even better, gifs) in the pr so we can see the functionality.

Added gifs to show bucket and object actions

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/bucketActions/CopyBucketAction.kt [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+        val bucketName = selected.toString()
+        val copyPasteManager = CopyPasteManager.getInstance()
+        copyPasteManager.setContents(StringSelection(bucketName))
+    }
+
+    override fun actionPerformed(selected: S3BucketNode, e: AnActionEvent) {
+        ApplicationManager.getApplication().executeOnPooledThread {
+            try {
+                performCopy(selected)
+                TelemetryService.getInstance().record(e.project, "s3") {
```
Comment (zhangzhx) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/bucketActions/CopyBucketAction.kt]: As the action extends `AnActionWrapper`, it already integrates the telemetry service. Do we still need to send the metric again here?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/UploadObjectAction.kt [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+import software.aws.toolkits.jetbrains.services.s3.bucketEditor.S3TreeTable
+import software.aws.toolkits.jetbrains.services.telemetry.TelemetryService
+import software.aws.toolkits.jetbrains.utils.notifyError
+import javax.swing.tree.DefaultMutableTreeNode
+
+class UploadObjectAction(
+    val bucket: S3VirtualBucket,
+    val treeTable: S3TreeTable,
+    private val fileChooserFactory: FileChooserFactory
+) : AnActionButton("Upload Object", null, AllIcons.Actions.Upload), TelemetryNamespace {
```
Comment (zhangzhx) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/UploadObjectAction.kt]: Why do we use `AnActionButton` here? We should use [ActionWrappers](https://github.com/aws/aws-toolkit-jetbrains/blob/master/jetbrains-core/src/software/aws/toolkits/jetbrains/components/telemetry/ActionWrappers.kt) for all our actions ideally.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/bucketActions/DeleteBucketAction.kt [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+
+import software.amazon.awssdk.services.s3.S3Client
+import software.aws.toolkits.core.s3.deleteBucketAndContents
+import software.aws.toolkits.jetbrains.core.AwsClientManager
+import software.aws.toolkits.jetbrains.core.DeleteResourceAction
+import software.aws.toolkits.jetbrains.services.s3.S3BucketNode
+import software.aws.toolkits.jetbrains.services.telemetry.TelemetryService
+import software.aws.toolkits.jetbrains.utils.TaggingResourceType
+
+class DeleteBucketAction : DeleteResourceAction<S3BucketNode>("Delete S3 Bucket", TaggingResourceType.S3_BUCKET) {
```
Comment (zhangzhx) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/bucketActions/DeleteBucketAction.kt]: Like Kyle pointed out, we should wrap all the action titles into `message`

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/bucketActions/S3BucketViewer.kt [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+import com.intellij.openapi.project.Project
+import software.amazon.awssdk.services.s3.S3Client
+import software.aws.toolkits.jetbrains.core.AwsClientManager
+import software.aws.toolkits.jetbrains.services.s3.S3BucketNode
+import software.aws.toolkits.jetbrains.services.s3.S3VirtualBucket
+import software.aws.toolkits.jetbrains.services.s3.S3VirtualFileSystem
+import software.aws.toolkits.jetbrains.services.telemetry.TelemetryService
+import software.aws.toolkits.jetbrains.utils.notifyError
+
+class S3BucketViewer : OpenBucketViewerAction<S3BucketNode>("Open Bucket in Viewer") {
```
Comment (zhangzhx) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/bucketActions/S3BucketViewer.kt]: What other action could potentially extends `OpenBucketViewerAction`?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/DeleteObjectAction.kt [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+import software.aws.toolkits.jetbrains.services.telemetry.TelemetryService
+import software.aws.toolkits.jetbrains.utils.notifyInfo
+import software.aws.toolkits.resources.message
+import javax.swing.tree.DefaultMutableTreeNode
+
+class DeleteObjectAction(
+    private var treeTable: S3TreeTable,
+    val bucket: S3VirtualBucket
+
+) : AnActionButton("Delete Object", null, AllIcons.Actions.Cancel), TelemetryNamespace {
```
Comment (zhangzhx) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/DeleteObjectAction.kt]: Remove the extra empty line.

Why do we use `AnActionButton` for all the object related actions? We should use [ActionWrappers](https://github.com/aws/aws-toolkit-jetbrains/blob/master/jetbrains-core/src/software/aws/toolkits/jetbrains/components/telemetry/ActionWrappers.kt) for all our actions ideally. I think it's better to add another wrapper for `AnActionButton` if we have to use it.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/DeleteObjectAction.kt [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+            val key = when (file.parent is S3VirtualDirectory) {
+                true -> "${file.parent.name}/${file.name}"
+                false -> file.name
+            }
+            objectsToDelete.add(ObjectIdentifier.builder().key(key).build())
+        }
+
+        val response = Messages.showOkCancelDialog(
+            project,
+            "Are you sure you want to delete ${rows.size} objects?",
```
Comment (zhangzhx) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/DeleteObjectAction.kt]: All these string literals should go to `message`

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/DeleteObjectAction.kt [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+            message("delete_resource.title", "Object", ""),
+            "Delete",
+            "Cancel", Messages.getWarningIcon()
+        )
+        if (response == 0) {
+            ApplicationManager.getApplication().executeOnPooledThread {
+                try {
+                    deleteObjectAction(client, objectsToDelete)
+                    TelemetryService.getInstance().record(e.project, "s3") {
```
Comment (zhangzhx) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/DeleteObjectAction.kt]: All these telemetry report should go to some action wrapper so we don't have to deal with it all the time.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/UploadObjectAction.kt [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+
+    @Suppress("unused")
+    override fun actionPerformed(e: AnActionEvent) {
+        val project = e.getRequiredData(LangDataKeys.PROJECT)
+        val client: S3Client = AwsClientManager.getInstance(project).getClient()
+        val descriptor = FileChooserDescriptorFactory.createMultipleFilesNoJarsDescriptor()
+            .withDescription("Upload object")
+
+        val row = treeTable.selectedRow
+        println(row)
```
Comment (zhangzhx) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/UploadObjectAction.kt]: Remove this.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/DeleteObjectAction.kt (hunk 11 above)
Comment (shruti0085) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/DeleteObjectAction.kt]: Editor popup menu wasn't able to register object actions with AnActionWrapper. I think I can try to create a separate wrapper for AnActionButton and have the telemetry reports in there.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/bucketActions/S3BucketViewer.kt (hunk 10 above)
Comment (shruti0085) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/bucketActions/S3BucketViewer.kt]: Makes sense. I only need one single class -`OpenBucketViewerAction`

**Comment by kiiadi** [comment]
Screenshot 2019-08-12 at 17 11 53
the "Open Bucket Viewer" action should be disabled until after indicies are built.

**Comment by abrooksv** [comment]
> Screenshot 2019-08-12 at 17 11 53
> 
> the "Open Bucket Viewer" action should be disabled until after indicies are built.

Should just be able to make it DumbAware to fix that

**Comment by kiiadi** [comment]
also looks like when you run the action from right-click you dont get the tab title, double-click works fine

File: jetbrains-core/resources/META-INF/plugin.xml [hunk 15]
```diff
@@ -221,6 +222,17 @@
                     class="software.aws.toolkits.jetbrains.services.cloudformation.DeleteCloudFormationStackAction"/>
         </group>
 
+        <group id="aws.toolkit.explorer.s3.bucket" popup="true" compact="false">
+            <action id="s3.open.bucket.viewer"
+                    class="software.aws.toolkits.jetbrains.services.s3.bucketActions.S3BucketViewer"/>
+            <action id="s3.create.bucket"
+                    class="software.aws.toolkits.jetbrains.services.s3.bucketActions.CreateS3Bucket"/>
```
[the remark this answers is not in the document]
Comment (kiiadi) [reply jetbrains-core/resources/META-INF/plugin.xml]: also create bucket should be attached to the service node (e.g S3) and not a bucket.

**Comment by kiiadi** [comment]
@aws/aws-ides-team - should each resource node in the explorer tree have a default 'open' action that works on double-click?

* Stack would be the Stack UI
* S3 bucket would be this viewer
* Lambda would be ... run?

**Comment by abrooksv** [comment]
> @aws/aws-ides-team - should each resource node in the explorer tree have a default 'open' action that works on double-click?
> 
>     * Stack would be the Stack UI
> 
>     * S3 bucket would be this viewer
> 
>     * Lambda would be ... run?

We had that for stack details, but its not a leaf node, so it also expanded the tree.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/DownloadObjectAction.kt [hunk 16]
```diff
@@ -15,35 +15,33 @@ import com.intellij.openapi.project.Project
...
+) : ActionButtonWrapper(message("s3.download.object.action"), null, AllIcons.Actions.Download) {
 
     constructor(treeTable: TreeTable, bucket: S3VirtualBucket) : this(treeTable, bucket, FileChooserFactory.getInstance())
 
     @Suppress("unused")
-
-    override fun actionPerformed(e: AnActionEvent) {
+    override fun doActionPerformed(e: AnActionEvent) {
         val project = e.getRequiredData(LangDataKeys.PROJECT)
         val client: S3Client = AwsClientManager.getInstance(project).getClient()
```
Comment (zhangzhx) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/DownloadObjectAction.kt]: Unlike the explorer node, the S3 client here might not be created correctly. As you open the bucket editor, you need to retain the credentials and region for that bucket into the editor since the active credentials and region could be changed outside. This client should be built from the retained credentials and region instead.

File: jetbrains-core/resources/META-INF/plugin.xml [hunk 17]
```diff
@@ -222,15 +222,18 @@
                     class="software.aws.toolkits.jetbrains.services.cloudformation.DeleteCloudFormationStackAction"/>
         </group>
 
+        <group id="aws.toolkit.explorer.s3" popup="false" compact="false">
```
Comment (zhangzhx) [inline jetbrains-core/resources/META-INF/plugin.xml]: I think it is better to use the original group id `aws.toolkit.explorer.s3.bucket`.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/DeleteObjectAction.kt [hunk 18]
```diff
@@ -13,22 +13,19 @@ import software.amazon.awssdk.services.s3.S3Client
...
 import software.aws.toolkits.jetbrains.utils.notifyInfo
 import software.aws.toolkits.resources.message
 import javax.swing.tree.DefaultMutableTreeNode
 
 class DeleteObjectAction(
     private var treeTable: S3TreeTable,
     val bucket: S3VirtualBucket
-
-) : AnActionButton("Delete Object", null, AllIcons.Actions.Cancel), TelemetryNamespace {
+) : AnActionButton(message("s3.delete.object.action"), null, AllIcons.Actions.Cancel) {
```
Comment (zhangzhx) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/DeleteObjectAction.kt]: Replace this with `AnActionButtonWrapper`?

Comment (shruti0085) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/DeleteObjectAction.kt]: Sorry. Missed it by mistake.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/DeleteObjectAction.kt [hunk 19]
```diff
@@ -75,10 +67,10 @@ class DeleteObjectAction(
     override fun isEnabled(): Boolean = (!(treeTable.isEmpty || (treeTable.selectedRow < 0) ||
             (treeTable.getValueAt(treeTable.selectedRow, 1) == "")))
 
-    override fun updateButton(e: AnActionEvent) {}
-
     override fun isDumbAware(): Boolean = true
 
+    override fun updateButton(e: AnActionEvent) { }
```
Comment (zhangzhx) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/DeleteObjectAction.kt]: I think you can move these two methods to the parent class.

File: jetbrains-core/resources/META-INF/plugin.xml (hunk 17 above)
Comment (shruti0085) [reply jetbrains-core/resources/META-INF/plugin.xml]: @kiiadi suggested that create bucket should be attached to the S3 service node.

Comment (zhangzhx) [reply jetbrains-core/resources/META-INF/plugin.xml]: Ah make sense.

**Comment by kiiadi** [comment]
Something weird has happened with this PR - is it rebased / merged correctly - I wouldn't expect there to be 58 commits.

**Comment by kiiadi** [comment]
> > @aws/aws-ides-team - should each resource node in the explorer tree have a default 'open' action that works on double-click?
> > ```
> > * Stack would be the Stack UI
> > 
> > * S3 bucket would be this viewer
> > 
> > * Lambda would be ... run?
> > ```
> 
> We had that for stack details, but its not a leaf node, so it also expanded the tree. You removed that, remember? 😛

oh yeah...

**Comment by shruti0085** [comment]
> Something weird has happened with this PR - is it rebased / merged correctly - I wouldn't expect there to be 58 commits.

Ideally, it should only be showing 8 commits. However, somehow its shows all the previous commits from the earlier PR as well. I can rebase by going back and redoing from the start but it might lead to loss of reviews and history. 
Austin suggested that since the earlier commits haven't changed and while merging all commits would be squashed, I can leave them for now.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/RenameObjectAction.kt [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+                }
+            }
+        }
+    }
+
+    override fun isEnabled(): Boolean = !(treeTable.isEmpty || (treeTable.selectedRow < 0) ||
+            (treeTable.getValueAt(treeTable.selectedRow, 1) == "") || (treeTable.selectedRows.size > 1))
+
+    @TestOnly
+    fun renameObjectAction(response: String, file: VirtualFile, client: S3Client) {
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/RenameObjectAction.kt]: why is this exposed as test only?

Comment (shruti0085) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/RenameObjectAction.kt]: Not required. Will fix it with the next PR.
